src/vdc_collection.rs

```diff
@@ -0,0 +1,245 @@
+        }
+    }
+}
+
+/// Verifiable Digital Credential Collection
+///
+/// This is the main interface to credentials.
+pub struct VdcCollection {
+    storage: Box<dyn StorageManagerInterface>,
+}
+
+#[derive(Error, Debug, uniffi::Error)]
+pub enum VdcCollectionError {
+    /// Attempt to convert the credential to a serialized form suitable for writing to storage failed.
+    #[error("Failed to Serialize Value")]
+    SerializeFailed,
+
+    /// Attempting to convert the credential to a deserialized form suitable for runtime use failed.
+    #[error("Failed to Deserialize Value")]
+    DeserializeFailed,
+
+    /// Attempting to write the credential to storage failed.
+    #[error("Failed to Write to Storage")]
+    StoreFailed(StorageManagerError),
+
+    /// Attempting to read the credential from storage failed.
+    #[error("Failed to Read from Storage")]
+    LoadFailed(StorageManagerError),
+
+    /// Attempting to delete a credential from storage failed.
+    #[error("Failed to Delete from Storage")]
+    DeleteFailed(StorageManagerError),
+}
+
+impl VdcCollection {
+    /// Create a new credential set.
+    pub fn new(engine: Box<dyn StorageManagerInterface>) -> VdcCollection {
+        VdcCollection { storage: engine }
+    }
+
+    /// Add a credential to the set.
+    pub fn add(
+        &self,
+        id: Uuid,
+        format: CredentialFormat,
+        ctype: CredentialType,
+        payload: Vec<u8>,
+    ) -> Result<(), VdcCollectionError> {
+        let val = match serde_cbor::to_vec(&Credential::new(id, format, ctype, payload)) {
+            Ok(x) => x,
+            Err(_) => return Err(VdcCollectionError::SerializeFailed),
+        };
+
+        match self.storage.add(self.id_to_key(id), Value(val)) {
+            Ok(()) => Ok(()),
+            Err(e) => Err(VdcCollectionError::StoreFailed(e)),
+        }
+    }
+
+    /// Get a credential from the store.
+    pub fn get(&self, id: &str) -> Result<Credential, VdcCollectionError> {
+        let raw = match self.storage.get(self.str_to_key(id)) {
+            Ok(x) => x.0,
+            Err(e) => return Err(VdcCollectionError::LoadFailed(e)),
+        };
+
+        match serde_cbor::de::from_slice(&raw) {
+            Ok(x) => Ok(x),
+            Err(_) => Err(VdcCollectionError::DeserializeFailed),
+        }
+    }
+
+    /// Remove a credential from the store.
+    pub fn delete(&self, id: &str) -> Result<(), VdcCollectionError> {
+        match self.storage.remove(self.str_to_key(id)) {
+            Ok(_) => Ok(()),
+            Err(e) => Err(VdcCollectionError::DeleteFailed(e)),
+        }
+    }
+
+    /// Get a list of all the credentials.
+    pub fn all_entries(&self) -> Vec<Key> {
+        let mut r = Vec::new();
+
+        for key in self.storage.list() {
+            let name = key.0;
+
+            if name.starts_with(KEY_PREFIX) {
+                if let Some(id) = name.strip_prefix(KEY_PREFIX) {
+                    r.push(Key(id.to_string()));
```

Comment:
Just return a string here, or add a [newtype](https://doc.rust-lang.org/rust-by-example/generics/new_types.html) for "credential id" here. `Key` from storage manager should be considered the path to the file. A consumer of this API should just care about the id of the credential.
```suggestion
                    r.push(id.to_string());
```

Reply:
Ok, I've made these changes.